core/project_settings.cpp

```diff
@@ -184,6 +184,7 @@ bool ProjectSettings::_set(const StringName &p_name, const Variant &p_value) {
 	// marking the project settings as dirty allows them only to be
 	// checked when dirty.
 	_dirty_this_frame = 2;
+	_version++;
```

Comment:
Wait, I totally missed this earlier.
`_version` absolutely must only be updated _after_ the full change to `ProjectSettings` has been registered.
Otherwise, the old value might be queried and cached for the new version number.

Reply:
Ah yes, I'll change it.

Reply:
Thanks, changed code looks safe 👍

Reply:
Well spotted. It's kind of a gotcha here because we don't really want to be checking another mutex for `_version` so just needs some care.